ext/pdo/pdo_stmt.c

```diff
@@ -1216,37 +1163,34 @@ PHP_METHOD(PDOStatement, fetchAll)
 	int flags = how & PDO_FETCH_FLAGS;
 
 	old_ce = stmt->fetch.cls.ce;
-	ZVAL_COPY_VALUE(&old_ctor_args, &stmt->fetch.cls.ctor_args);
-	old_arg_count = stmt->fetch.cls.fci.param_count;
-
-	do_fetch_opt_finish(stmt, 0);
+	old_ctor_args = stmt->fetch.cls.ctor_args;
 
 	/* TODO Would be good to reuse part of pdo_stmt_setup_fetch_mode() in some way */
-
 	switch (fetch_mode) {
 		case PDO_FETCH_CLASS:
 			/* Figure out correct class */
 			if (arg2) {
 				if (Z_TYPE_P(arg2) != IS_STRING) {
 					zend_argument_type_error(2, "must be of type string, %s given", zend_zval_value_name(arg2));
-					RETURN_THROWS();
+					goto reset_old_ctor_args;
 				}
 				stmt->fetch.cls.ce = zend_lookup_class(Z_STR_P(arg2));
 				if (!stmt->fetch.cls.ce) {
 					zend_argument_type_error(2, "must be a valid class");
-					RETURN_THROWS();
+					goto reset_old_ctor_args;
 				}
 			} else {
 				stmt->fetch.cls.ce = zend_standard_class_def;
 			}
 
-			if (ctor_args && zend_hash_num_elements(Z_ARRVAL_P(ctor_args)) > 0) {
-				ZVAL_COPY_VALUE(&stmt->fetch.cls.ctor_args, ctor_args); /* we're not going to free these */
-			} else {
-				ZVAL_UNDEF(&stmt->fetch.cls.ctor_args);
+			if (ctor_args && zend_hash_num_elements(ctor_args) > 0) {
+				if (!stmt->fetch.cls.ce->constructor) {
+					zend_argument_value_error(3, "must be empty when class provided in argument #2 ($class) does not have a constructor");
+					goto reset_old_ctor_args;
```

Comment:
I have a similar concern as above: you jump to `reset_old_ctor_args` which will also release `ctor_args`, but it was not increased in refcount yet.

Reply:
Actually, I wonder why you increment the refcount anyway, the old code didn't do it?